contracts/fate.md

```diff
@@ -0,0 +1,820 @@
+* EC, Execution continuation: A list of instructions left in the current basic block.
+
+### Memory
+The machine memory is divided into:
+* The chain top (block height, etc)
+* Event stream
+* The account state tree
+* The contract store
+* The execution/code memory
+* Execution stack (push/pop on call/return)
+* Accumulator stack (push/pop on instruction level)
+* Local storage/stack/environment/context (push/pop on let...in...end)
+
+#### The execution memory
+
+The code to execute is stored in the execution memory. The execution
+memory is a three level map. The key on the top level is a contract
+address. The second level is a map from function hashes to a map of
+basic blocks. The keys in the map of basic blocks are indices to basic
+blocks (a 0 based, dense enumeration). The values are lists of
+instructions.
+
+When a contract is called the code of the contract is read from the
+state tree into the execution memory. At this point the machine
+implementation can deserialize the serialized version of FATE code
+into instructions and basic blocks.
+
+The serialization format for FATE code is described in Appendix 1:
+[Fate instruction set and serialization](#appendix-1-fate-serialization).
+
+#### The chain
+
+The chain "memory" contains information about the current block and
+the current iteration. These values are available through special
+instructions.
+
+* Beneficiary
+* Timestamp
+* (Block) Height
+* Difficulty
+* Gaslimit
+* Address
+* Balance
+* Caller
+* Value
+
+#### The contract store
+
+The permanent storage of a contract. Stored in the chain state
+tree. This is a key value store that is mapped to the contract state
+by the compiler.
+
+The compiler can choose how to represent the contract state in the
+store. For instance, save the entire state under a single key (this is
+how it works in AEVM at the moment), or it could flatten any state
+records and store each field under a separate key.
+
+#### The state tree
+
+Contains all accounts, oracles and contracts on the chain. Most values
+such as account balances can be read directly through specific load
+instructions. Some values can be changed through transaction
+instructions.
+
+#### The local storage
+
+The local storage is a key value store.
+The key is a an integer. The value can be of any FATE type.
+The local store is local to a function call.
+
+#### The accumulator stack
+
+The accumulator stack is a stack of typed values. The top of the stack
+can be used as an argument to an operation. Values can be pushed to
+the stack and popped from the stack.
+
+#### The execution stack
+
+The execution stack contain return addresses as a tuple in the form of
+{contract address, function hash, and basic block index}.
+
+#### The event stream
+
+A contract can emit events similar to the EVM.
+
+## Types
+
+The machine supports the following types:
+* Integer (signed arbitrary precision integers)
+* Boolean (true | false)
+* Strings (arbitrary size byte arrays)
```

Comment:
String?

Reply:
also, should the text in the parentheses be plural or singular? Similar issues below.